src/build.cc

```diff
@@ -768,32 +774,40 @@ bool Builder::FinishCommand(CommandRunner::Result* result, string* err) {
     }
 
     if (node_cleaned) {
-      TimeStamp restat_mtime = 0;
-      // If any output was cleaned, find the most recent mtime of any
-      // (existing) non-order-only input or the depfile.
-      for (vector<Node*>::iterator i = edge->inputs_.begin();
-           i != edge->inputs_.end() - edge->order_only_deps_; ++i) {
-        TimeStamp input_mtime = disk_interface_->Stat((*i)->path(), err);
-        if (input_mtime == -1)
-          return false;
-        if (input_mtime > restat_mtime)
-          restat_mtime = input_mtime;
-      }
-
-      string depfile = edge->GetUnescapedDepfile();
-      if (restat_mtime != 0 && deps_type.empty() && !depfile.empty()) {
-        TimeStamp depfile_mtime = disk_interface_->Stat(depfile, err);
-        if (depfile_mtime == -1)
-          return false;
-        if (depfile_mtime > restat_mtime)
-          restat_mtime = depfile_mtime;
-      }
-
       // The total number of edges in the plan may have changed as a result
       // of a restat.
       status_->PlanHasTotalEdges(plan_.command_edge_count());
+    }
 
-      output_mtime = restat_mtime;
+    // Use the time from the most recent input that was computed when the edge was
+    // started, not the mtime of the edge's latest output as it is now. There could have
+    // been other edges that restat'd the input node and detected a change, but for *this*
+    // edge, we want the mtime as it was when the command began.
+    //
+    // Generator edges should just use the edge's latest output mtime since it's possible that
+    // the generator rule could create/update implicit inputs as part of the command, and
+    // we don't want to get into an infinite loop.
+    if (!edge->GetBindingBool("generator")) {
+      record_mtime = edge->most_recent_input_mtime_;
+
+      // If there were any added deps, compute the most recent input mtime
+      for (vector<Node*>::iterator i = deps_nodes.begin();
+          i != deps_nodes.end(); ++i) {
+        (*i)->StatIfNecessary(disk_interface_, err);
+        if ((*i)->mtime() > record_mtime)
+          record_mtime = (*i)->mtime();
+      }
+    }
+
+    // Take into account the mtime of the depfile. Special depfiles (denoted with a "deps"
+    // binding) are handled above
```

Comment:
The "Special depfiles" are just depfiles that are loaded into the DepsLog. Perhaps:

```c++
// If we have 'depfile' but not 'deps' then the depfile is not
// loaded into the deps log.  Take its mtime into account.
```